test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/MaxRequestBodySizeTests.cs

```diff
@@ -0,0 +1,384 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+using Microsoft.AspNetCore.Testing;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.FunctionalTests
+{
+    public class MaxRequestBodySizeTests
+    {
+        [Fact]
+        public async Task RejectsRequestWithContentLengthHeaderExceedingGlobalLimit()
+        {
+            // 4 GiB
+            var globalMaxRequestBodySize = 0x100000000;
+            BadHttpRequestException requestRejectedEx = null;
+
+            using (var server = new TestServer(async context =>
+            {
+                var buffer = new byte[1];
+                requestRejectedEx = await Assert.ThrowsAsync<BadHttpRequestException>(
+                    async () => await context.Request.Body.ReadAsync(buffer, 0, 1));
+                throw requestRejectedEx;
+            },
+            new TestServiceContext { ServerOptions = { Limits = { MaxRequestBodySize = globalMaxRequestBodySize } } }))
+            {
+                using (var connection = server.CreateConnection())
+                {
+                    await connection.Send(
+                        "POST / HTTP/1.1",
+                        "Host:",
+                        "Content-Length: " + (globalMaxRequestBodySize + 1),
+                        "",
+                        "");
+                    await connection.ReceiveForcedEnd(
+                        "HTTP/1.1 413 Payload Too Large",
```

Comment:
Mostly correct. All future reads would also fail, so even if the app swallowed all exceptions and didn't write to the response body, the request draining code would see the exception which would still trigger a 413. If the app writes to the response body before the request is drained (or it's a non-keepalive request that doesn't get drained), then there might not be a 413 response.